### ISSUE TITLE
dbstructure update error 1265: Data truncated for column 'thumb'

### ISSUE THREAD

**Issue by Alkarex** [friendica/friendica#4666]
In addition to https://github.com/friendica/friendica/issues/4665 , I have an error during dbstructure update:

```
php bin/console.php dbstructure update
ALTER TABLE `contact` DROP INDEX `uid_rel_network_poll`, DROP INDEX `uid_network_batch`, MODIFY `id` int unsigned NOT NULL auto_increment COMMENT '', MODIFY `uid` mediumint unsigned NOT NULL DEFAULT 0 COMMENT 'User id', MODIFY `rel` tinyint unsigned NOT NULL DEFAULT 0 COMMENT '', MODIFY `network` char(4) NOT NULL DEFAULT '' COMMENT '', MODIFY `photo` varchar(255) NOT NULL DEFAULT '' COMMENT '', MODIFY `thumb` varchar(255) NOT NULL DEFAULT '' COMMENT '', MODIFY `micro` varchar(255) NOT NULL DEFAULT '' COMMENT '', MODIFY `request` varchar(255) COMMENT '', MODIFY `notify` varchar(255) COMMENT '', MODIFY `poll` varchar(255) COMMENT '', MODIFY `confirm` varchar(255) COMMENT '', MODIFY `poco` varchar(255) COMMENT '', MODIFY `priority` tinyint unsigned NOT NULL DEFAULT 0 COMMENT '', MODIFY `contact-type` tinyint NOT NULL DEFAULT 0 COMMENT '', MODIFY `rating` tinyint NOT NULL DEFAULT 0 COMMENT '', MODIFY `closeness` tinyint unsigned NOT NULL DEFAULT 99 COMMENT '', MODIFY `profile-id` int unsigned NOT NULL DEFAULT 0 COMMENT '', MODIFY `fetch_further_information` tinyint unsigned NOT NULL DEFAULT 0 COMMENT '', ADD INDEX `uid_rel_network_poll` (`uid`,`rel`,`network`,`poll`(64),`archive`), ADD INDEX `uid_network_batch` (`uid`,`network`,`batch`(64)), COMMENT = 'contact table';

Error 1265 occurred during database update:
Data truncated for column 'thumb' at row 393
Errors encountered performing database changes: ALTER TABLE `contact` DROP INDEX `uid_rel_network_poll`, DROP INDEX `uid_network_batch`, MODIFY `id` int unsigned NOT NULL auto_increment COMMENT '', MODIFY `uid` mediumint unsigned NOT NULL DEFAULT 0 COMMENT 'User id', MODIFY `rel` tinyint unsigned NOT NULL DEFAULT 0 COMMENT '', MODIFY `network` char(4) NOT NULL DEFAULT '' COMMENT '', MODIFY `photo` varchar(255) NOT NULL DEFAULT '' COMMENT '', MODIFY `thumb` varchar(255) NOT NULL DEFAULT '' COMMENT '', MODIFY `micro` varchar(255) NOT NULL DEFAULT '' COMMENT '', MODIFY `request` varchar(255) COMMENT '', MODIFY `notify` varchar(255) COMMENT '', MODIFY `poll` varchar(255) COMMENT '', MODIFY `confirm` varchar(255) COMMENT '', MODIFY `poco` varchar(255) COMMENT '', MODIFY `priority` tinyint unsigned NOT NULL DEFAULT 0 COMMENT '', MODIFY `contact-type` tinyint NOT NULL DEFAULT 0 COMMENT '', MODIFY `rating` tinyint NOT NULL DEFAULT 0 COMMENT '', MODIFY `closeness` tinyint unsigned NOT NULL DEFAULT 99 COMMENT '', MODIFY `profile-id` int unsigned NOT NULL DEFAULT 0 COMMENT '', MODIFY `fetch_further_information` tinyint unsigned NOT NULL DEFAULT 0 COMMENT '', ADD INDEX `uid_rel_network_poll` (`uid`,`rel`,`network`,`poll`(64),`archive`), ADD INDEX `uid_network_batch` (`uid`,`network`,`batch`(64)), COMMENT = 'contact table'; <br />
```

Updating from Friendica develop 06d6d6c758b26713f97e3f9929f39e496171a023 to the current /develop.

On Ubuntu 17.10, with default MySQL 5.7.21-0ubuntu0.17.10.1.

**Comment by Alkarex** [comment]
It looks like NULL values in contact.thumb where not taken into account:

```sql
ALTER TABLE `contact` MODIFY `thumb` varchar(255) NOT NULL DEFAULT '';
ERROR 1265 (01000): Data truncated for column 'thumb' at row 393

SELECT id, addr, alias, name, nick, thumb from contact WHERE thumb IS NULL;
+--------+-------------------------------+--------------------------------+------------------+-----------------+-------+
| id     | addr                          | alias                          | name             | nick            | thumb |
+--------+-------------------------------+--------------------------------+------------------+-----------------+-------+
|    411 | ffeth@framasphere.org         |                                | Feth Arezki      | ffeth           | NULL  |
|   2818 | Stirhok@mamot.fr              | https://mamot.fr/users/Stirhok | Amiral Poulet Ξ  | Stirhok         | NULL  |
|  16818 | badumtss@framapiaf.org        |                                | badumtss         | badumtss        | NULL  |
| 363815 | fourelus@mamot.fr             |                                | Quentin          | fourelus        | NULL  |
| 366628 | max_inredcastle@framapiaf.org |                                | MaximG           | max_inredcastle | NULL  |
+--------+-------------------------------+--------------------------------+------------------+-----------------+-------+
5 rows in set (0.03 sec)

UPDATE contact SET thumb = '' WHERE thumb IS NULL;

ALTER TABLE `contact` MODIFY `thumb` varchar(255) NOT NULL DEFAULT '';
Query OK, 3859 rows affected (4.72 sec)
```

**Comment by Alkarex** [comment]
One more column was wrong. Fixed with:

```sql
UPDATE contact SET micro = '' WHERE micro IS NULL;
```

**Comment by annando** [comment]
That's weird. These are several fields that are created as "not null" that **are** null ...

**Comment by Alkarex** [comment]
Well, the type of those fields used to allow NULL:
```sql
`thumb` text,
`micro` text,
...
```
https://github.com/friendica/friendica/blob/e12b6e01a23ede9e7e134bc4615d14dc618ebebd/database.sql#L107-L128

**Comment by annando** [comment]
f*** ... now I understand. We changed these fields to ```varchar(255) NOT NULL DEFAULT ''``` in the new version.

**Comment by MrPetovan** [comment]
Do we have a way to apply pre-update scripts?

**Comment by Alkarex** [comment]
After the fix, e.g. `UPDATE contact SET micro = '' WHERE micro IS NULL;` then the `ALTER TABLE` command needs to be called again.

**Comment by MrPetovan** [comment]
That's why I'm asking about a pre-update script because we already have a post-update process but it wouldn't apply here.

**Comment by annando** [comment]
The easiest way is to change the database structure so that all fields that formally had been "text" and now are "varchar" don't get a "not null".

**Comment by MrPetovan** [comment]
3.6.1 confirmed?

**Comment by annando** [comment]
What does @tobiasd mean? I could work on a nice solution this day. Tomorrow I will be busy, since there is a handball game.

**Comment by tobiasd** [comment]
I wont have time for a new release for a week or so. Even with the steps now being a bit more clear. Also I think there are to many open problems at the moment.

**Comment by MrPetovan** [comment]
Closed through #4678
